proposals/proposals/nnnn-trailing-comma-tuples-arguments-conditions.md

```diff
@@ -0,0 +1,200 @@
+
+```swift
+extension [S] {
+    func sorted<each T: Comparable>(_ keyPath: repeat KeyPath<S, each T>) { }
+}
+
+arrayOfS.sorted(
+  \.a, 
+  \.b,
+//  \.c
+) ❌ Unexpected ',' separator
+```
+
+Since [#21381](https://github.com/apple/swift/pull/21381) has been merged back in 2019 **enum associated values** supports default values and are a good fit for trailing comma as well.
+
+
+**Tuples** use are very close to arguments list and, although may not be so frequently used, it seems natural that they adopt trailing comma too.
+
+**Multiple conditions** in `if`, `guard` and `while` are also comma-separated list and add, remove, reorder and comment in/out are not uncommon practice during development.
+
+```swift
+if 
+   condition1,
+   condition2,
+//   condition3
+{ ❌ Cannot convert value of type '() -> ()' to expected condition type 'Bool'
+                    
+} ❌ Expected '{' after 'if' condition
+```
+### Code Generation
+
+**Plugins** and **Macros** have made it possible to generate code using swift and trailing comma would allow generate list of arguments and conditions without worrying about a special condition for the last element.
+
+### Code Diff
+
+A tangential motivation is that trailing comma makes version-control diffs cleaner.
+
+Without trailing comma:
+```diff
+foo(
+-  a: Int
++  a: Int,
++  b: Int
+)
+```
+With trailing comma:
+```diff
+foo(
+  a: Int,
++  b: Int,
+)
+```
+
+> [!NOTE]  
+> A similar proposal was [rejected](https://forums.swift.org/t/rejected-se-0084-allow-trailing-commas-in-parameter-lists-and-tuples/2777) back in 2016 for Swift 3. It's been 8 years since that, the swift language has evolved a lot, some changes highlighted above as motivation, and the code style that "puts the terminating right parenthesis on a line following the arguments to that call" has been widely adopted by community, swift standard library codebase, swift-format, docc documentation and Xcode. Therefore, not encourage or endorse this code style doesn't hold true anymore nor is a reason for rejection.
+
+## Proposed solution
+
+This proposal adds support for trailing comma to:
+
+### Arguments and Parameters
+
+Including declaration and call of initializers, functions and enum case associated values.
+
+```swift
+
+func foo(
+    a: Int = 0, 
+    b: Int = 0, 
+) {
+}
+
+foo(
+    a: 1,
+    b: 2,
+)
+
+```
+
+### Tuples
+
+```swift
+(1, 2, 3,)
+```
+
+### Conditions
+
+Including `if`, `guard` and `while`.
+
+```swift
+if 
+   condition1,
+   condition2,
+   condition3,
+{
+                    
+}
+```
+
+## Source compatibility
+
+This change is purely additive and has no impact on existing code.
+
+## Future directions
+
+### Allow trailing comma anywhere there's a comma-separated list
+
+Although this proposal focuses on the most requested use cases for trailing comma, there's other places with comma-separated list and the restriction could be consistently lifted for all of these.
```

Comment:
Missing from this list are variable bindings, e.g., `var a, b, c`.

Also worth calling out explicitly (in the proposed solution) is whether this proposal will support `var (a, b,) = (1, 2,)`, or whether only the right-hand side is allowed to have a trailing comma—in other words, whether a tuple pattern counts as a "tuple" under the proposed solution.